process/resultsParser/resultsParser.go

```diff
@@ -0,0 +1,401 @@
+	logger "github.com/multiversx/mx-chain-logger-go"
+)
+
+var log = logger.GetOrCreate("api/gin")
+
+// ResultOutcome encapsulates data contained within the smart contact results.
+type ResultOutcome struct {
+	ReturnCode    ReturnCode      `json:"returnCode"`
+	ReturnMessage string          `json:"returnMessage"`
+	Values        []*bytes.Buffer `json:"values"`
+}
+
+// ParseResultOutcome will try to translate the smart contract results into a ResultOutcome object.
+func ParseResultOutcome(tx *transaction.ApiTransactionResult, pubKeyConverter core.PubkeyConverter) (*ResultOutcome, error) {
+	outcome := parseOutcomeOnSimpleMoveBalance(tx)
+	if outcome != nil {
+		log.Trace("txHash [%s] result outcome on simple move balance", tx.Hash)
+		return outcome, nil
+	}
+
+	outcome = parseOutcomeOnInvalidTransaction(tx)
+	if outcome != nil {
+		log.Trace("txHash [%s] result outcome on invalid transaction", tx.Hash)
+		return outcome, nil
+	}
+
+	outcome, err := parseOutcomeOnEasilyFoundResultWithReturnData(tx.SmartContractResults)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse outcome on easily found result with return data: %w", err)
+	}
+	if outcome != nil {
+		log.Trace("txHash [%s] result outcome on easily found result with return data", tx.Hash)
+		return outcome, nil
+	}
+
+	outcome, err = parseOutcomeOnSignalError(tx.Logs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse outcome on signal error: %w", err)
+	}
+	if outcome != nil {
+		log.Trace("txHash [%s] result outcome on signal error", tx.Hash)
+		return outcome, nil
+	}
+
+	outcome, err = parseOutcomeOnTooMuchGasWarning(tx.Logs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse outcome on too much gas warning: %w", err)
+	}
+	if outcome != nil {
+		log.Trace("txHash [%s] result outcome on too much gas warning", tx.Hash)
+		return outcome, nil
+	}
+
+	outcome, err = parseOutcomeOnWriteLogWhereFirstTopicEqualsAddress(tx.Logs, tx.Sender)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse outcome on write log where first topic equals address: %w", err)
+	}
+	if outcome != nil {
+		log.Trace("txHash [%s] result outcome on write log where first topic equals address", tx.Hash)
+		return outcome, nil
+	}
+
+	outcome, err = parseOutcomeWithFallbackHeuristics(tx, tx.Receivers[0])
+	if outcome != nil {
+		log.Trace("txHash [%s] result outcome on fallback heuristics", tx.Hash)
+		return outcome, nil
+	}
+	return nil, nil
+}
+
+func parseOutcomeOnSimpleMoveBalance(tx *transaction.ApiTransactionResult) *ResultOutcome {
+	noResults := len(tx.SmartContractResults) == 0
+	var noLogs bool
+	if tx.Logs != nil {
+		noLogs = len(tx.Logs.Events) == 0
+	} else {
+		noLogs = true
+	}
+
+	if noResults && noLogs {
+		return &ResultOutcome{
+			ReturnCode:    None,
+			ReturnMessage: None.String(),
+		}
+	}
+
+	return nil
+}
+
+func parseOutcomeOnInvalidTransaction(tx *transaction.ApiTransactionResult) *ResultOutcome {
+	if tx.Status == transaction.TxStatusInvalid {
```

Comment:
you should refactor this function and use early exit instead of imbricated if statements

Reply:
refactored.

Reply:
not done